File: rektofun-program/programs/rektofun-program/src/instructions/initialize_config.rs
use anchor_lang::prelude::*;

use crate::{constants::*, state::Config};

/// One-time bootstrap: creates the global `Config` PDA, seeded with the
/// compile-time defaults from `constants.rs`. Only the genesis `ADMIN_PUBKEY`
/// can call this — after it succeeds, `Config::admin` (mutable via
/// `update_admin`) becomes the live source of admin authority everywhere else.
#[derive(Accounts)]
pub struct InitializeConfig<'info> {
    #[account(mut, address = ADMIN_PUBKEY)]
    pub admin: Signer<'info>,

    #[account(
        init,
        payer = admin,
        space = 8 + Config::INIT_SPACE,
        seeds = [CONFIG_SEED],
        bump,
    )]
    pub config: Account<'info, Config>,

    pub system_program: Program<'info, System>,
}

pub(crate) fn handler(ctx: Context<InitializeConfig>) -> Result<()> {
    let config = &mut ctx.accounts.config;
    config.admin = ADMIN_PUBKEY;
    config.platform_fee_bps = PLATFORM_FEE_BPS;
    config.creator_fee_bps = CREATOR_FEE_BPS;
    config.min_bet_amount = MIN_BET_AMOUNT;
    config.max_duration_secs = MAX_DURATION_SECS;
    config.min_duration_secs = MIN_DURATION_SECS;
    config.max_team_size = MAX_TEAM_SIZE;
    config.bump = ctx.bumps.config;

    msg!("Config initialized — admin: {}", config.admin);

    Ok(())
}

File: rektofun-program/scripts/initialize-config.ts
/**
 * One-time bootstrap script: calls `initialize_config` to create the global
 * `Config` PDA on-chain, seeded with the compile-time defaults from
 * `src/constants.rs` (platform_fee_bps, creator_fee_bps, min_bet_amount, etc).
 *
 * Only the genesis `ADMIN_PUBKEY` (see src/constants.rs) can sign this call —
 * the same secret used by `app/lib/admin-signer.ts` server-side.
 *
 * Usage:
 *   ADMIN_PRIVATE_KEY=<base58-secret> yarn ts-node scripts/initialize-config.ts
 *
 * Optional env vars:
 *   RPC_URL     - defaults to Anchor.toml's devnet cluster
 */
import { AnchorProvider, Program, Idl } from "@anchor-lang/core";
import { Connection, Keypair, PublicKey, SystemProgram } from "@solana/web3.js";
import bs58 from "bs58";
import idl from "../target/idl/rektofun_program.json";

const RPC_URL = process.env.RPC_URL || "https://api.devnet.solana.com";
const CONFIG_SEED = Buffer.from("config");
const PROGRAM_ID = new PublicKey(idl.address);

async function main() {
  const privateKeyBase58 = process.env.ADMIN_PRIVATE_KEY;
  if (!privateKeyBase58) {
    throw new Error(
      "Set ADMIN_PRIVATE_KEY (base58 secret for the genesis admin wallet, " +
        "same as used by app/lib/admin-signer.ts) to run this script."
    );
  }
  const admin = Keypair.fromSecretKey(bs58.decode(privateKeyBase58));
  const connection = new Connection(RPC_URL, "confirmed");
  const wallet = {
    publicKey: admin.publicKey,
    signTransaction: async (tx: any) => {
      tx.partialSign(admin);
      return tx;
    },
    signAllTransactions: async (txs: any[]) => {
      txs.forEach((tx) => tx.partialSign(admin));
      return txs;
    },
  };
  const provider = new AnchorProvider(connection, wallet as any, {
    commitment: "confirmed",
  });
  const program = new Program(idl as Idl, provider);

  const [configPda, bump] = PublicKey.findProgramAddressSync(
    [CONFIG_SEED],
    PROGRAM_ID
  );

  console.log("Admin:", admin.publicKey.toBase58());
  console.log("Config PDA:", configPda.toBase58(), "bump:", bump);

  const existing = await connection.getAccountInfo(configPda);
  if (existing) {
    console.log("Config already initialized — nothing to do.");
    return;
  }

  const sig = await (program.methods as any)
    .initializeConfig()
    .accounts({
      admin: admin.publicKey,
      config: configPda,
      systemProgram: SystemProgram.programId,
    })
    .rpc();

  console.log("initialize_config confirmed:", sig);
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
